aten/src/ATen/core/optional.h

```diff
@@ -301,6 +301,14 @@ using OptionalBase = typename std::conditional<
 
 template <class T>
 class optional : private OptionalBase<T> {
+
+  template <class U>  // re-declaration for nvcc on Windows.
```

Comment:
I wish we knew what actual MSVC/NVCC bug we were triggering here, but it's not very obvious.

Reply:
Will do!

Reply:
From messing around on the windows machine it's something like conditional template specifications aren't visible inside a class that extends a conditional template specification.  When I defined TensorBase as either of the conditional arguments it compiled fine, but as soon as the conditional was introduced it broke (even if the if was just true or false), even when I defined my own conditional.